kimchi/src/snarky/boolean.rs

```diff
@@ -0,0 +1,192 @@
+use ark_ff::PrimeField;
+
+use crate::{
+    loc,
+    snarky::{
+        checked_runner::{RunState, TypeCreation},
+        constraint_system::BasicSnarkyConstraint,
+        cvar::CVar,
+        traits::SnarkyType,
+    },
+};
+
+trait OutOfCircuitSnarkyType2<F> {
+    type InCircuit;
+}
+
+impl<F> OutOfCircuitSnarkyType2<F> for bool
+where
+    F: PrimeField,
+{
+    type InCircuit = Boolean<F>;
+}
+
+/// A boolean variable.
+#[derive(Clone)]
+pub struct Boolean<F: PrimeField>(CVar<F>);
+
+impl<F> SnarkyType<F> for Boolean<F>
+where
+    F: PrimeField,
+{
+    type Auxiliary = ();
+
+    type OutOfCircuit = bool;
+
+    const SIZE_IN_FIELD_ELEMENTS: usize = 1;
+
+    fn to_cvars(&self) -> (Vec<CVar<F>>, Self::Auxiliary) {
+        (vec![self.0.clone()], ())
+    }
+
+    fn from_cvars_unsafe(cvars: Vec<CVar<F>>, _aux: Self::Auxiliary) -> Self {
+        assert_eq!(cvars.len(), Self::SIZE_IN_FIELD_ELEMENTS);
+        Self(cvars[0].clone())
+    }
+
+    fn check(&self, cs: &mut RunState<F>) {
+        // TODO: annotation?
+        cs.assert_(None, vec![BasicSnarkyConstraint::Boolean(self.0.clone())]);
+    }
+
+    fn deserialize(&self) -> (Self::OutOfCircuit, Self::Auxiliary) {
+        todo!()
+    }
+
+    fn serialize(out_of_circuit: Self::OutOfCircuit, aux: Self::Auxiliary) -> Self {
+        todo!()
+    }
+
+    fn constraint_system_auxiliary() -> Self::Auxiliary {
+        todo!()
+    }
+
+    fn value_to_field_elements(x: &Self::OutOfCircuit) -> (Vec<F>, Self::Auxiliary) {
+        todo!()
+    }
+
+    fn value_of_field_elements(x: (Vec<F>, Self::Auxiliary)) -> Self::OutOfCircuit {
+        todo!()
+    }
+}
+
+impl<F> Boolean<F>
+where
+    F: PrimeField,
+{
+    pub fn true_() -> Self {
+        Self(CVar::Constant(F::one()))
+    }
+
+    pub fn false_() -> Self {
+        Self(CVar::Constant(F::zero()))
+    }
+
+    pub fn not(&self) -> Self {
+        Self(Self::true_().0 - &self.0)
+    }
+
+    pub fn if_(&self, then_: Self, else_: Self) -> Self {
+        todo!()
```

Comment:
It would be nice to remove this and use the 'generic' `if_`. Something like:
```
pub fn if_<F: PrimeField, T: SnarkyType<F>>(cs: RunState<F>, b: Boolean<F>, then_: T, else_: T) -> T
```
which should be easy-ish to implement using `to_cvars` and `of_cvars_unsafe` (I believe there's already a version of this in the snarky repo.)